Dpll

melanie-taprogge opened the pull request:
This is a first draft for the doll implementation!

Commit:
some tests

Commit:
import of dimacs lightly implemented

Commit:
fixed sets

Commit:
dimacsinput completed. YEAH

Commit:
completed input

Commit:
little adjustment

qualmluc committed:
Merge branch 'main' into IO-modules

Commit:
fixed format problem

Commit:
fixed everything mentioned in the meeting

Commit:
...and adjusted readme

Commit:
fixed cambridge problems

melanie-taprogge committed:
added a first draft of the dpll implementation

dpll as well as supporting functions (like conditioning, selecting the most common literal, ...)

Luc4sQ commented:
looks great so far, but i saw you returning like

```
return satisfiability, variableAssignment
```

if you use instead a tuple

```
return (satisfiability, variableAssignment)
```

it is more inline with our code and you can unpack them simply when accessing the method

```
true, assignment = dp.dpll(someKNF)
```

i will review the rest in depth later, but clean code tho

qualmluc commented on src/alg/dpll.py:
```diff
@@ -0,0 +1,103 @@
+    knf_variables_list = np.array([])
+    for array in cnf:
+        for entry in array:
+            knf_variables_list = np.append(knf_variables_list,abs(entry))
```
I'm not sure, if this is unnecessary pushing the runtime, of the DPLL or if the specific choice is compensating for this, but we should take care of this on the long run, i think

melanie-taprogge replied:
I do think that we will probably use more advanced methods of choosing the next variable to try in the long run, I see this rather as a starting point as well and figured this would probably be better than choosing a variable completely at random! :)

qualmluc commented on src/alg/dpll.py:
```diff
@@ -0,0 +1,103 @@
+    # we do this recursively by repeatedly calling dpll_alg and assigning variables through conditioning
+    # until one of the first two cases applies and the cnf is therefore either sat or unsat.
+    else:
+        # we first select the variable for conditioning 
```
here you could make it a little more readable if you substitute the nested "Else" statements into one larger "elif" statement imo.

qualmluc replied:
to explain further what i mean

```
if statement:
    ...
else:
    if statement2:
        ...
    else:
        ...
```

is equivalent to

```
if statement:
    ...
elif statement2:
    ...
else:
    ...
```

melanie-taprogge replied:
I do agree that it would be nicer but there is actually a reason why I didn't end up doing it that way:
If I would get rid of the nested if-else statement, I would have to run the calculations that take place after the first "else" and before the nested "if" (I am talking about "var_candidate = findMostCommonVar(cnf)" and "new_cnf = conditioning(cnf,var_candidate)") every time the function "doll_alg" is called because would have to place them before the first if-condition (since they can't be run in between else-statements). So the calculations would be done every time the function is iteratively called and not only when the first two conditions are not satisfied (i.e. when the cnf is either empty or contains an empty clause). Running the calculations is however unnecessary in situations where one of these cases applies and would therefore unnecessarily increase computing time! So I decided to go with the nested variation for the sake of an improved runtime even though its not quite as nice!
Hope you understand what I mean!

qualmluc replied:
thats not true actually. you can write code like

```
if something:
    ...
elif emptythingy
    ...
elif dpll_alg(conditioning(cnf,findMostCommonVar(cnf))) != "unsat":
            var.append(var_candidate)
            return(var)
elif rest?
    ...
```
that code wouldn't run, when the statements above in DPLL are true, so the runtime concerns wouldn't be necessary.

melanie-taprogge replied:
That would work but I think I would still have to either of the two following things when it comes to the calculation of the new variable:
- define it at the beginning of the dpll_alg (i.e. before the if-else conditions) and therefore calculate the new variable (findMostCommonVar(cnf)) either every time dpll_alg is called even if one of the first two conditions is satisfied and it isn't necessary
- not define the new variable (or define it after already running the function that chooses it in order to determine weather we are in a case where we need to add it to the list of variables) but instead calculate it every time it is either needed in the condition or appended to the list of variables. But That way I would also run the operation multiple times instead of just calling the function that chooses the next variable one.

Or am I overseeing something? ;)

melanie-taprogge replied:
Alo I thought that readability might benefit from defining the variables first as "new_cnf" and "var_candidate" rather than just using the functions that calculate them directly in the conditions. I figured this way it might be a bit easier to follow what the operations actually do but if you guys think that this is superfluous I can for sure avoid defining "new_cnf" by just using the function "conditioning" directly in the elif or if statement the way you proposed @Luc4sQ ! :)

qualmluc replied:
i've looked a little more deep into that and you're right. I overlooked this detail. that would end up in unnecessary calls of functions, as i don't know if python supports variable definition in a function call. (I think i saw that somewhere, but im not sure)

if you prefer your style in terms of readability, that is totally fine for me :) With my critic i was primarily aiming for sharing my opinion. ultimately codingstyle depends on taste

qualmluc reviewed (approved):
All in all, very nice implementation of basic dpll (Algorithm 3.3 i guess?)

I tried to run it with a benchmark file which has 50 variables, what isn't thinkable with brute force and it finished with the right output! I guess thats a serious success

melanie-taprogge committed:
changed dpll output to return a tuple

Thank you @Luc4sQ for pointing it out!

melanie-taprogge committed:
updated variable assignment output

I realised that I listed the "system internal" notation of negated variables  in the output variable assignment, which also resulted in entries like "{..., -7: False, ...}". (Which is unnecessary since the assignment to False makes the "-" notation superfluous) I changed it so that instead it returns the absolute value of the variable!
